### PR TITLE
iox-#17: Fix cargo build

### LINKED ISSUE #17: cargo build error
## Required information

**Operating system:**
E.g. Ubuntu 20.04 LTS

**Rust version:**
E.g. Rust 1.55.0

**Observed result or behaviour:**
A clear and precise description of the observed result.

**Expected result or behaviour:**
What do you expect to happen?

**Conditions where it occurred / Performed steps:**
Describe how one can reproduce the bug.

`root@MSI:~/2.iceoryx-rs/iceoryx-rs# cargo build
   Compiling iceoryx-rs v0.0.10 (/root/2.iceoryx-rs/iceoryx-rs)
The following warnings were emitted during compilation:

warning: src/runtime.rs:8:14: fatal error: iceoryx_posh/runtime/posh_runtime.hpp: No such file or directory
warning:     8 |     #include "iceoryx_posh/runtime/posh_runtime.hpp"
warning:       |              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: compilation terminated.

error: failed to run custom build command for `iceoryx-rs v0.0.10 (/root/2.iceoryx-rs/iceoryx-rs)`

Caused by:
  process didn't exit successfully: `/root/2.iceoryx-rs/iceoryx-rs/target/debug/build/iceoryx-rs-3895b6884be6cb5f/build-script-build` (exit status: 1)
  --- stdout
  Output { status: ExitStatus(unix_wait_status(0)), stdout: "", stderr: "HEAD is now at ebdfcc5df Merge pull request #1312 from ApexAI/iox-#1311-increase-version-to-2.0.1\n" }
  Output { status: ExitStatus(unix_wait_status(0)), stdout: "", stderr: "" }
  Output { status: ExitStatus(unix_wait_status(0)), stdout: "-- Configuring done\n-- Generating done\n-- Build files have been written to: /root/2.iceoryx-rs/iceoryx-rs/target/iceoryx-build/iceoryx_hoofs\n", stderr: "" }
  Output { status: ExitStatus(unix_wait_status(512)), stdout: "[ 17%] Built target iceoryx_platform\n[ 20%] Building CXX object CMakeFiles/iceoryx_hoofs.dir/source/posix_wrapper/access_control.cpp.o\n", stderr: "In file included from /root/2.iceoryx-rs/iceoryx-rs/target/iceoryx-git/iceoryx_hoofs/include/iceoryx_hoofs/internal/posix_wrapper/access_control.hpp:23,\n                 from /root/2.iceoryx-rs/iceoryx-rs/target/iceoryx-git/iceoryx_hoofs/source/posix_wrapper/access_control.cpp:18:\n/root/2.iceoryx-rs/iceoryx-rs/target/iceoryx-git/iceoryx_hoofs/platform/linux/include/iceoryx_hoofs/platform/acl.hpp:20:10: fatal error: sys/acl.h: No such file or directory\n   20 | #include <sys/acl.h>\n      |          ^~~~~~~~~~~\ncompilation terminated.\nmake[2]: *** [CMakeFiles/iceoryx_hoofs.dir/build.make:271: CMakeFiles/iceoryx_hoofs.dir/source/posix_wrapper/access_control.cpp.o] Error 1\nmake[1]: *** [CMakeFiles/Makefile2:96: CMakeFiles/iceoryx_hoofs.dir/all] Error 2\nmake: *** [Makefile:130: all] Error 2\n" }
  Output { status: ExitStatus(unix_wait_status(0)), stdout: "", stderr: "" }
  Output { status: ExitStatus(unix_wait_status(256)), stdout: "-- Configuring incomplete, errors occurred!\nSee also \"/root/2.iceoryx-rs/iceoryx-rs/target/iceoryx-build/iceoryx_posh/CMakeFiles/CMakeOutput.log\".\nSee also \"/root/2.iceoryx-rs/iceoryx-rs/target/iceoryx-build/iceoryx_posh/CMakeFiles/CMakeError.log\".\n", stderr: "CMake Error at CMakeLists.txt:24 (find_package):\n  By not providing \"Findiceoryx_hoofs.cmake\" in CMAKE_MODULE_PATH this\n  project has asked CMake to find a package configuration file provided by\n  \"iceoryx_hoofs\", but CMake did not find one.\n\n  Could not find a package configuration file provided by \"iceoryx_hoofs\"\n  with any of the following names:\n\n    iceoryx_hoofsConfig.cmake\n    iceoryx_hoofs-config.cmake\n\n  Add the installation prefix of \"iceoryx_hoofs\" to CMAKE_PREFIX_PATH or set\n  \"iceoryx_hoofs_DIR\" to a directory containing one of the above files.  If\n  \"iceoryx_hoofs\" provides a separate development package or SDK, be sure it\n  has been installed.\n\n\n" }
  Output { status: ExitStatus(unix_wait_status(512)), stdout: "", stderr: "make: *** No rule to make target 'install'.  Stop.\n" }
  TARGET = Some("x86_64-unknown-linux-gnu")
  OPT_LEVEL = Some("0")
  HOST = Some("x86_64-unknown-linux-gnu")
  CXX_x86_64-unknown-linux-gnu = None
  CXX_x86_64_unknown_linux_gnu = None
  HOST_CXX = None
  CXX = None
  CXXFLAGS_x86_64-unknown-linux-gnu = None
  CXXFLAGS_x86_64_unknown_linux_gnu = None
  HOST_CXXFLAGS = None
  CXXFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("true")
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")
  running: "c++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-I" "/root/2.iceoryx-rs/iceoryx-rs" "-I" "/root/2.iceoryx-rs/iceoryx-rs/target/iceoryx-install/include/iceoryx/v2.0.1" "-Wall" "-Wextra" "-Wno-noexcept-type" "-std=c++14" "-o" "/root/2.iceoryx-rs/iceoryx-rs/target/debug/build/iceoryx-rs-e6a6ce3817c727f2/out/rust_cpp/cpp_closures.o" "-c" "/root/2.iceoryx-rs/iceoryx-rs/target/debug/build/iceoryx-rs-e6a6ce3817c727f2/out/rust_cpp/cpp_closures.cpp"
  cargo:warning=src/runtime.rs:8:14: fatal error: iceoryx_posh/runtime/posh_runtime.hpp: No such file or directory
  cargo:warning=    8 |     #include "iceoryx_posh/runtime/posh_runtime.hpp"
  cargo:warning=      |              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=compilation terminated.
  exit status: 1

  --- stderr

  error occurred: ToolExecError: Command "c++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-I" "/root/2.iceoryx-rs/iceoryx-rs" "-I" "/root/2.iceoryx-rs/iceoryx-rs/target/iceoryx-install/include/iceoryx/v2.0.1" "-Wall" "-Wextra" "-Wno-noexcept-type" "-std=c++14" "-o" "/root/2.iceoryx-rs/iceoryx-rs/target/debug/build/iceoryx-rs-e6a6ce3817c727f2/out/rust_cpp/cpp_closures.o" "-c" "/root/2.iceoryx-rs/iceoryx-rs/target/debug/build/iceoryx-rs-e6a6ce3817c727f2/out/rust_cpp/cpp_closures.cpp" with args "c++" did not execute successfully (status code exit status: 1).`

### PR DESCRIPTION
**Pull request by adoerr**
Fixes #17 

The build issue is not related to `libacl1-dev`.

The problem actually seems to be that the current `make_and_install()` function is not dealing with the `cpptoml` dependency of `iceoryx_posh` correctly.

This in turn leads to a failure with the `cmake` configuration of `iceoryx_posh`. You can check this by inspecting the `./target/iceoryx-install/include/*` folders. Basically the `iceoryx_posh` include folder is missing, leading to the error message referenced in the issue.

The PR essentially just mimics the recommended way to build `iceoryx` itself. Admittedly this is not ideal in terms of minimal compilation. On the other hand, it seems to work reliably. Please, feel free to optimize!

I have also taken the liberty to bump `ICEORYX_VERSION` to `v2.0.2`.

## Pre-Review Checklist for the PR Author

1. [x ] Code follows the Rust coding style and is formatted with `rustfmt`
1. [x ] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [ ] Commits messages are according to this [guideline][commit-guidelines]
    - [ ] Commit messages have the issue ID (`iox-#42 commit text`)
    - [ ] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x ] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x ] Relevant issues are linked
1. [x ] Add sensible notes for the reviewer
1. [x ] All checks have passed (except `task-list-completed`)
1. [ x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php

## Checklist for the PR Reviewer

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code according to our coding style and naming conventions
- [ ] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## References

- Closes **TBD**

### CODE REVIEW

**Comment by dkroenke** [comment]
@adoerr For the `eclipsefdn/eca` check you need  to sign the (ECA)[https://www.eclipse.org/legal/ECA.php] in your Eclipse Account with the E-Mail address that you have used for signing the commits.

File: src/sb/ffi.rs [hunk 1]
```diff
@@ -266,7 +266,7 @@ impl Subscriber {
 
 impl fmt::Debug for Subscriber {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "{:#?}", self)
+        f.debug_struct("Subscriber").field("_opaque", &self._opaque).finish()
```
Comment (dkroenke) [inline src/sb/ffi.rs:269]: Just some formatting, otherwise looks good!

```suggestion
        f.debug_struct("Subscriber")
            .field("_opaque", &self._opaque)
            .finish()
```

File: build.rs [hunk 2]
```diff
@@ -7,46 +7,24 @@ use std::path::Path;
...
-                out
-            })
-            .map(|out| println!("{:?}", out))?;
-    }
+    Command::new("cmake")
+        .current_dir(source_dir)
+        .arg(format!("-B{}", build_dir))
+        .arg("-Hiceoryx_meta")
+        .arg("-DCMAKE_BUILD_TYPE=Release")
+        .arg("-DBUILD_SHARED_LIBS=OFF")
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply build.rs:15]: I'm not sure about this change. It adds some overhead and it should be possible to build the libs separately like we do it in ROS.

I can reproduce the error from #17 when I remove libacl from my system, even with your changes. So the problem you have might be something else. 

I also noticed that I used `std::process::Command` in the wrong way since I did not check for the status.

Could you use this code and share the log of a failed build
```rs
use std::io::{Error, ErrorKind};
fn make_and_install(source_dir: &str, build_dir: &str, install_dir: &str) -> std::io::Result<()> {
    let cmake_install_prefix = format!("-DCMAKE_INSTALL_PREFIX={}", install_dir);

    for iceoryx_component in &["iceoryx_hoofs", "iceoryx_posh"] {
        let component_source_dir = format!("{}/{}", source_dir, iceoryx_component);
        let component_build_dir = format!("{}/{}", build_dir, iceoryx_component);

        if !Command::new("mkdir")
            .args(&["-p", &component_build_dir])
            .status()?
            .success()
        {
            return Err(Error::new(
                ErrorKind::Other,
                format!("Could not create build dir for '{}'!", iceoryx_component),
            ));
        }

        if !Command::new("cmake")
            .current_dir(&component_build_dir)
            .args(&[
                "-DCMAKE_BUILD_TYPE=Release",
                "-DBUILD_SHARED_LIBS=OFF",
                &cmake_install_prefix,
                &component_source_dir,
            ])
            .status()?
            .success()
        {
            return Err(Error::new(
                ErrorKind::Other,
                format!("Could not run cmake for '{}'!", iceoryx_component),
            ));
        }

        if !Command::new("cmake")
            .current_dir(&component_build_dir)
            .args(&["--build", ".", "--target", "install"])
            .status()?
            .success()
        {
            return Err(Error::new(
                ErrorKind::Other,
                format!("Could not build '{}'!", iceoryx_component),
            ));
        }
    }

    Ok(())
}
```

File: examples/publisher_simple.rs [hunk 3]
```diff
@@ -14,7 +14,7 @@ struct CounterTopic {
     counter: u32,
 }
 
-unsafe impl POD for CounterTopic {}
+impl POD for CounterTopic {}
```
Comment (elBoberido) [inline examples/publisher_simple.rs:17]: Not sure about this either. I know that it is not nice to have `unsafe` in the code but this marker trait is similar to `Send` and `Sync`. One has to be careful to implement it since the data which can be send via shared memory has some constraints. I think I could use the `Copy` trait as bound but that would restrict the data types more than necessary. Ideally there would be a `CouldBeCopy` trait bound. I would be happy about a solution without `unsafe` but currently one has to be careful when the trait is implemented. Ideas how to solve this in a different way are welcome and highly appreciated.

**Comment by elBoberido** [comment]
@adoerr #21 will abort the build process early if iceoryx cannot be build. With that we should be able to find the root cause of your issue.

I also added a naive clippy fix but I like your one more, so please don't dismiss it when you rebase onto master

File: build.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (adoerr) [reply build.rs:15]: Please see the `build.log` attached to #21

File: examples/publisher_simple.rs (hunk 3 above)
Comment (adoerr) [reply examples/publisher_simple.rs:17]: I think right now, this is mostly a documentation issue. `clippy` was complaining about [missing safety doc](clippy::missing_safety_doc). I did not see a need for `POD`  to be `unsafe`, so I removed the qualifier :smile: 

The most elegant solution is probably having some sort of a `iox::pod` attribute macro. This macro would check the constraints and  generate the code for shared memory decoding / encoding and the Trait implementations needed.

Something like

```Rust
#[iox::pod]
pub struct CounterTopic {
...
}
```

So in other words, the idea behind `iox::pod` would be to hide as much of the `iceoryx` protocol / machinery as possible.

**Comment by adoerr** [comment]
@elBoberido thanks for looking into this!

This PR is basically just a quick-and-dirty solution that allowed me to compile `iceoryx-rs` somehow. So it is probably best to close this PR and continue with #21.

File: examples/publisher_simple.rs (hunk 3 above)
Comment (elBoberido) [reply examples/publisher_simple.rs:17]: @adoerr yes, that would be ideal. If you have the knowledge, time and inclination it would be a great contribution. My knowledge regarding proc macros is quite limited but it is on my todo list :smile:

**Comment by elBoberido** [comment]
Closed in favor of #21. Relevant commits are cherry-picked to that PR
